### PR TITLE
Symbolic computation of Jacobian of QSS mechanisms

### PR DESCRIPTION
**Pull request by malihass**
1)Use `sympy` and `symengine` to symbolically encode the production rates.
2)Allow the user to print symbolic Jacobian in place of the old Jacobian. Effect is shown below. 
Left was old Jacobian implementations (works for very small timesteps). Right is the new one.

aj_0D_QSS

This should also help address the GMRES instabilities that we noticed. Below is what happens with finite difference Jac, analytical Jac and GMRES for the same 0D test case with same tolerances and timestepping

qss_integrator

3)Print the Jacobian using common sub-expression eliminations
4)Do some management of common subexpression to reduce the memory footprint. Format options can be controlled by the user via `.toml` file
5) Update documentation to explain how this procedure works.
6) The `mechanism.H` for `dodecane_lu_qss` is the one with symbolically computed Jacobian.

Co-authored-by: Nicholas Wimer [nicholas.wimer@gmail.com](mailto:nicholas.wimer@gmail.com)

### CODE REVIEW

File: Support/ceptr/makeQss_toml.sh [hunk 1]
```diff
@@ -0,0 +1,3 @@
+poetry run qssa -f ${PELE_PHYSICS_HOME}/Support/Mechanism/Models/dodecane_lu_qss/skeletal.yaml -n ${PELE_PHYSICS_HOME}/Support/Mechanism/Models/dodecane_lu_qss/non_qssa_list.yaml
```
Comment (marchdf) [inline Support/ceptr/makeQss_toml.sh]: remove this file and just put the example commands in the documentation.

File: Support/ceptr/qssa_input.toml [hunk 2]
```diff
@@ -0,0 +1,23 @@
+[Readability]
```
Comment (marchdf) [inline Support/ceptr/qssa_input.toml]: Remove this file and just put it in the documentation.

File: Support/ceptr/ceptr/species_info.py [hunk 3]
```diff
@@ -61,3 +63,254 @@ def __init__(self):
...
+        numb_list = []
+        symb_list = []
+
+        for idx in range(self.n_species):
+            name_list.append(f"sc[{idx}]")
+            numb_list.append(idx)
+            symb_list.append(self.nonqssa_species_list[idx])
+
+        # Initialize the dataframe
+        sc_df = pd.DataFrame(
```
Comment (marchdf) [inline Support/ceptr/ceptr/species_info.py:267]: Can this be a plain `OrderedDict`? That way we don't have to import/install/depend on `pandas`. Just a thought...

File: Support/ceptr/ceptr/ceptr.py [hunk 4]
```diff
@@ -1,45 +1,91 @@
...
+        type=str,
+        default=None,
+        required=False,
+    )
+
+    parser.add_argument(
+        "-sj",
+        "--symbolic_jacobian",
+        action="store_true",
+        help="Compute the Jacobian using symbolic recording",
```
Comment (marchdf) [inline Support/ceptr/ceptr/ceptr.py]: Is this only for QSS? Maybe add that to the help string? Maybe the option should be `qss_symbolic_jacobian`?

File: Support/ceptr/ceptr/ceptr.py [hunk 5]
```diff
@@ -1,45 +1,91 @@
...
+        description="Mechanism converter", formatter_class=RawTextHelpFormatter
+    )
     group = parser.add_mutually_exclusive_group(required=True)
     group.add_argument("-f", "--fname", help="Mechanism file", type=str)
     group.add_argument(
         "-l", "--lst", help="Mechanism directory file list", type=str
     )
+
+    parser.add_argument(
+        "--format_input",
```
Comment (marchdf) [inline Support/ceptr/ceptr/ceptr.py]: can we call this option `qss_jacobian_format` ?

File: Support/ceptr/ceptr/converter.py [hunk 6]
```diff
@@ -2,17 +2,20 @@
 import pathlib
 import shutil
 import subprocess as spr
+import time
```
Comment (marchdf) [inline Support/ceptr/ceptr/converter.py]: Remove all `time` stuff. We can just use a profiler.

File: Support/ceptr/ceptr/species_info.py (hunk 3 above)
Comment (marchdf) [reply Support/ceptr/ceptr/species_info.py:267]: Chatted with Nick and I am convinced that pandas should stay.

**Comment by esclapez** [comment]
@malihass I'll give it a try on a case today. Can't say much about the python :)

**Review by marchdf — APPROVED** [review]

**Review by esclapez — APPROVED** [review]
Successfully tested the dodecane_qss mechanism on an single igniting dodecane jet, and the Jacobian seems to work nicely !

**Pull request merged**